### PR TITLE
Lasso Instability Fix

### PR DESCRIPTION
**Pull request by owerbat**
A couple of new functions for writing stable algorithms have been added:
* daal::parallel_deterministic_reduce - tbb::parallel_deterministic_reduce wrapper
* daal::parallel_deterministic_sum - special case when result of reduce is an array

New functionality has been used for stabilizing of Lasso algorithm

### CODE REVIEW

**Comment by owerbat** [comment]
make test

File: algorithms/threading/threading.h [hunk 1]
```diff
@@ -318,6 +326,46 @@ inline bool is_in_parallel()
...
+}
+
+template <typename F1, typename F2, typename F3, typename F4>
+inline void * parallel_deterministic_reduce(size_t n, size_t grain_size, const F1 & init_function, const F2 & delete_function,
+                                            const F3 & loop_function, const F4 & reduce_function)
+{
+    const void * a = static_cast<const void *>(&init_function);
+    const void * b = static_cast<const void *>(&delete_function);
+    const void * c = static_cast<const void *>(&loop_function);
+    const void * d = static_cast<const void *>(&reduce_function);
```
Comment (averbukh) [inline algorithms/threading/threading.h]: const void * const ... = ...

File: algorithms/threading/threading.cpp [hunk 2]
```diff
@@ -633,5 +633,71 @@ DAAL_EXPORT void _daal_wait_task_group(void * taskGroupPtr) {}
...
+    for (int i = 0; i < n; ++i)
+    {
+        loop_func(local, i, i + 1, c);
+        reduce_func(total, local, d);
+    }
+
+    delete_func(&local, b);
+#endif
+
+    return total;
```
Comment (gyrlanda) [inline algorithms/threading/threading.cpp:699]: Would this "total" pointer look on previously deleted memory?

**Review by ntfshard — COMMENTED** [review]
I'm not a part of this project and this PR seems outdated.

**Pull request closed without merging**